engine/access/rpc/client/grpc_secure_client.go

```diff
@@ -0,0 +1,106 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/hex"
+	"fmt"
+	"log"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	libp2ptls "github.com/libp2p/go-libp2p-tls"
+	"github.com/onflow/flow/protobuf/go/flow/access"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/network/p2p"
+)
+
+func main() {
+
+	accessNodeAddress := "localhost:9000"
+
+	bytes, err := hex.DecodeString("691ef1777846fb6e0ad15d5dce77c10617acafab16cf6635df4a50dd29e6e9d448e22a0eee587904165153a8e5a44df92fec898a2134b2efe663a691b4bbcbe8")
+	if err != nil {
+		log.Fatalf("failed to decode public key: %v", err)
+	}
+	publicFlowNetworkingKey, err := crypto.DecodePublicKey(crypto.ECDSAP256, bytes)
+	if err != nil {
+		log.Fatalf("failed to create the Flow networking public key: %v", err)
+	}
+	fmt.Printf(" using public key %s for the remote node\n", publicFlowNetworkingKey.String())
+
+	publicLibp2pNetworkingKey, err := p2p.PublicKey(publicFlowNetworkingKey)
+	if err != nil {
+		log.Fatalf("failed to generate a libp2p key from a Flow key: %v", err)
+	}
+	remotePeerLibP2PID, err := peer.IDFromPublicKey(publicLibp2pNetworkingKey)
+	if err != nil {
+		log.Fatalf("failed to derive the libp2p Peer ID from the libp2p public key: %v", err)
+	}
+
+	tlsConfig, err := ConfigForPeer(remotePeerLibP2PID)
+	if err != nil {
+		log.Fatalf("failed to derive the TLS Config for the remote peer: %v", err)
+	}
+
+	conn, err := grpc.Dial(accessNodeAddress, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+	if err != nil {
+		log.Fatalf("did not connect: %v", err)
+	}
+	defer conn.Close()
+
+	// create an AccessAPIClient
+	grpcClient := access.NewAccessAPIClient(conn)
+
+	ctx := context.Background()
+
+	request := access.GetLatestBlockRequest{
+	}
+
+	response, err := grpcClient.GetLatestBlock(ctx, &request)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println("latest block height: ")
+	fmt.Println(response.GetBlock().String())
+}
+
+func ConfigForPeer(remote peer.ID) (*tls.Config, error) {
+
+	config := &tls.Config{
+		MinVersion:               tls.VersionTLS13,
+		InsecureSkipVerify:       true, // This is not insecure here. We will verify the cert chain ourselves.
```

Comment:
this is set to true since we are not validating hostnames from the certificate

Reply:
So, while I understand we're not interested in checking this certificate against a certificate chain:
- this is still a *self*-signed certificate,
- which constitutes a proof of possession,
- and so I would want to verify that signature.

Reply:
I think you want something close to the following for self-signed verification:

```
  config := &tls.Config {
                  ...
                 // Use the pool in the verify options:
		ClientAuth:               tls.RequireAnyClientCert,
                 VerifyConnection: func(cs tls.ConnectionState) error {
			opts := x509.VerifyOptions{
			    // create a new pool
			    Roots: x509.NewCertPool(),
			}
			// add the certificate itself as signer
			opts.Roots.AddCert(cs.PeerCertificates[0])
			_, err := cs.PeerCertificates[0].Verify(opts)
			return err
		},
	        ...
```

Reply:
The `ClientAuth` is used on the server side only 

https://pkg.go.dev/crypto/tls#Config

> 	// ClientAuth determines the server's policy for
> 	// TLS Client Authentication. The default is NoClientCert.
> 	ClientAuth ClientAuthType

Reply:
and wouldn't that  ☝️  `VerifyConnection` basically validate the cert against itself (since there is no chain here)?

Reply:
> and wouldn't that ☝️ VerifyConnection basically validate the cert against itself (since there is no chain here)?

Absolutely, that's precisely the intent.

> The ClientAuth is used on the server side only

Yes, sorry, same logic without it.
See `ExampleConfig_verifyConnection` in https://golang.org/src/crypto/tls/example_test.go